### DISCUSSION TITLE
Stride 4.2 / .NET Release (November 2023)

### DISCUSSION THREAD

**Discussion by VaclavElias** [stride3d/stride#1699] [Contributors]
Hello Striders,

Let's initiate a dialogue on an important topic. I've recently found myself, somewhat unintentionally, tasked with updating our websites. With the upcoming release of our new version, I want us to be prepared and poised for success. If feasible, we could coordinate to release **4.2** just a few days post the .NET 8 release, which would be an excellent achievement.

I've observed some substantial updates and clean-up work ongoing in the Stride repository. In light of this, it is crucial that we also keep our website and communication channels current and reflective of these changes.

Here's a brief rundown of what we have planned:

- The primary Stride website will feature an announcement for the 4.2 release, spotlighting recent fixes, enhancements, and updates. We can also take this opportunity to highlight any noteworthy updates or fixes from the recent 4.1 updates which we didn't communicate.
- The Stride Docs website will be updated with the latest Release Notes and other pages, such as the Manual, as necessary.
- The Stride Wiki, where we currently house contributor instructions, will be updated to reflect any changes relevant to our contributors.

I also had a discussion with xen2, and we're contemplating a 4.2 RC release once there is .NET 8 RC release available.

In addition, I'd like to propose the following tasks for us to consider:

- Review and possibly enhance our Stride game examples
- Update the manual if necessary.

I invite you all to contribute your ideas and suggestions. **Please consider what we could realistically accomplish given our current development capacity**. Please share any existing GitHub Issues you're currently working on and believe should be completed for the .NET 8 release. Post them in the chat so I can compile them into this section, which will facilitate easier tracking of these tasks.

- [x] https://github.com/stride3d/stride/pull/1123
- [x] https://github.com/stride3d/stride/pull/1158
- [ ] https://github.com/stride3d/stride/issues/1176
- [x] https://github.com/stride3d/stride/issues/1650
- [x] https://github.com/stride3d/stride/issues/1652
- [x] https://github.com/stride3d/stride/issues/1703
  - [x] https://github.com/stride3d/stride/pull/1704
- [x] https://github.com/stride3d/stride/pull/1717
- [x] https://github.com/stride3d/stride/pull/1719
- [x] https://github.com/stride3d/stride/issues/1750
- [x] https://github.com/stride3d/gettextnet/pull/2
- [x] https://github.com/stride3d/stride/pull/1827
- [x] https://github.com/stride3d/stride/pull/1821
- [x]  https://github.com/stride3d/stride/pull/1801
- [x] https://github.com/stride3d/stride/pull/1864
- [x] https://github.com/stride3d/stride/pull/1921
- [x] https://github.com/stride3d/stride/pull/1977
- [x] https://github.com/stride3d/stride/pull/2000
- [x] https://github.com/stride3d/stride/pull/2049
- [x] https://github.com/stride3d/stride/pull/2038
- [x] https://github.com/stride3d/stride/issues/2043
- [x] https://github.com/stride3d/stride-website/pull/235
- [x] https://github.com/stride3d/stride/pull/2064
- [x] https://github.com/stride3d/stride/pull/2077

Many more bug fixes but these are some new features/QOL changes

Your insights and input are invaluable to the growth of our project. Let's continue to strive for the best as we move forward.

I'll be actively involved in updating the website, documents, and manual. As I am still familiarizing myself with the game engine, I am unable to generate content independently at this point. Therefore, I greatly value and encourage your input. Please don't hesitate to draft any documentation; I'm more than willing to take your contributions, refine them, and bring them to completion. Your expertise and assistance in this process is greatly appreciated.

Shall we start? 🚀

**Comment by NicusorN5** [comment]
In my own opinion, the master branch is currently unusable because of https://github.com/stride3d/stride/issues/1692 and https://github.com/stride3d/stride/issues/1700 . There's also an other major bug related to the UI page editor that I will soon create an GitHub issue for (namely: text labels and buttons have invisible text with size 0). ~~If I had enough experience with Stride's internal codebase, I'd ifrst try to fix these problems, then move on to writing features. I suspect these issues were caused by the recent "cleanup" commits :/~~ (Nevermind, I have seen that .NET 8 release date is in November, so I expect the recent major issues to be fixed by then)

Also, are there other major features for Stride 4.2 except upgrading to .NET 8?

~~I don't think this is a great idea, atleast for the very near future.~~

__________________________________________________________________________________________________________

**Reply by VaclavElias** [reply]
Sorry for confusing you with the dates. I have updated the titles so we know it is most likely the November 2023. I also hope that any major bug will be fixed by then.

I am not aware about any major features done unless someone will correct me.

These are our official bounties for the new features https://opencollective.com/stride3d/projects. I am not sure if someone did something else major or minor outside of that. Let's see when others join the conversation here 🙂

**Comment by Ethereal77** [comment]
I'd also like that release to be a move to .NET 8 in as many parts of the engine as possible (and that includes replacing `xamarin-andrioid` and `xamarin-ios` with their most modern `net8.0-android` counterparts).
Also, I'd like to finish the project started several months ago of replacing SharpDX once and for all with the more modern Silk.DirectX, including using modern .NET primitives and all. That's the direction I'm moving to with my simplification and test PRs.
Finally, the other PR that tries to replace the C++ hand-written and old wrapper around Assimp with a full C# importer that uses Silk.Assimp I think would allow us to better ponder possible fixes, enhancements, etc, for the mesh import pipeline.

I think these three things are complex but achievable before november 2023. Or so I hope.

**Reply by VaclavElias** [reply]
Thank you @Ethereal77. If we have issues created for those you mentioned, could you provide it here so we can track it in this discussion? I will add it to the top. I believe you are getting assistance but if you think somebody could help just let us know.

**Reply by Ethereal77** [reply]
With the Android and iOS parts, I can start them later, but other people should do those parts because that's not my area and also I have no Apple devices.
For the Assimp one, there's [this PR](https://github.com/stride3d/stride/pull/1158) started by [johang88](https://github.com/johang88) that should serve as the base to complete the work. There's no associated issue, but I'll create it later if no one does so we can track it.
For the SharpDX replacement, [there's also a PR](https://github.com/stride3d/stride/pull/1123) by [ykafia](https://github.com/ykafia) and an issue #1176 that I plan to take into account in the following weeks if all goes well.

**Reply by VaclavElias** [reply]
We can track actually also through existing PRs, so the issue is not needed.

**Comment by Ethereal77** [comment]
Also I had an idea the other day that we should make samples, but not samples like the current templates that can be instantiated in the editor. I mean standalone samples using just code to program against the Stride framework to show basic things that can be made with Stride.
I'm talking things like the [client-server sample](https://github.com/xen2/Xenko.ClientServerSample) and [code-only samples](https://github.com/xen2/Xenko.CodeOnlySample) that [xen2](https://github.com/xen2) made, but also integrations with UI frameworks (WPF, Avalonia, WinForms, etc.). Simple things that show how to use the API without all the complexity of a full game project, scripts, or entities and components.
Other projects like MonoGame (and the old XNA) thrived with a community that made these simple samples that showed the potential the framework had.

If this is an interesting idea, maybe its own directory apart from `samples` (or within, but separated), or even its own repository `stride-samples` or something of that sort would be nice.

This is not excluding the creation of new template-samples like the ones we have now, of course. And also note these would be "official" or Stride-promoted samples. Not intended to compete with the samples or prototypes the community has made some of which are very good.

Also, these samples may serve as a way to test ideas that can later be promoted to templates, or that can help shape the API for enhancements or conveniences.

**Reply by VaclavElias** [reply]
I think this is a great idea regarding the content, however not sure if we have got enough capacity for this for this release. I myself would like to contribute with code only samples (https://github.com/VaclavElias/stride-code-only) as I find it very helpful for the beginners. I am currently working on new docs, which I hope will be ready for the 4.2 release, maybe also for 4.1. Then I would like to review the manual and actually learn about Stride Game engine, so I am going to have lots of input for docs/manual improvements and along the way I will be creating/testing such samples. Unless somebody else start working on this what you suggested I can possible join the effort after the 4.2 release.

**Comment by Doprez** [comment]
There is a very easy fix for some of the beginner samples that will fix the following:
https://github.com/stride3d/stride/issues/1652
https://github.com/stride3d/stride/issues/1650
this issue is currently fixed in the master branch but I dont think the main build pipeline updates the samples so new users are still fighting with it.

The .NET 8 PR was done here https://github.com/stride3d/stride/pull/1616 but it was an old preview so Im not 100% if it needs to be updated.

The Assimp silk.net PR works I am currently using it in my personal custom branch. The only issue is when I tried to use the same MeshConverter class for FBX files it breaks when a model has multiple meshes.

I think on our next release blog we should try to mention the community libraries. A lot of my issues from when I started with Stride were basic things that were not in engine but other members of Stride have made a library for. examples:
[awesome-stride I spent a lot of time searching for cool repos to gather this all together](https://github.com/Doprez/Awesome-Stride)
Useful extensions managed by the community
[Doprez.Stride](https://github.com/Doprez/Doprez.Stride)
[NexCoreExtension](https://github.com/NexStandard/NexCoreExtension)
[StrideToolkit - outdated but works perfectly with a simple nuget update](https://github.com/dfkeenan/StrideToolkit)
There are a bunch of things the community has made that seemed to be forgotten and they help/solve a lot of problems I see in Discord and Issues board.

**Reply by Ethereal77** [reply]
I've already fixed those related to the beginner samples in a PR I'll submit tomorrow in which I try to make the test projects compile and pass. There were some more problems with some of the samples. Nothing complex however.

(one of the test projects checks if the samples run and even tests little interactions to see if they respond well)

**Reply by Ethereal77** [reply]
Also, do some of those convenience extensions make sense to include in the Stride framework? If some of those things are not too specific and don't polute the API, maybe some can be included?

Btw, good work with [awesome Stride](https://github.com/Doprez/Awesome-Stride).

**Reply by Doprez** [reply]
Thanks! and I have brought up a few of the API changes before but it seems like they aren't interested in putting them directly in engine. I do think having them would be very useful for new Stride devs though, a lot of basic things like world position and delta time are awkward to access.

**Reply by VaclavElias** [reply]
I mentioned it here and there that I would like to create an official Stride Community Toolkit, where we could pack all our extensions and helpers and who knows what which are not yet in the Stride framework so we can evaluate it and see the impact. Once certain parts become essential, we can propose or advocate to move them to the Stride framework. Creating such a toolkit should be an example also how to document everything so it is easy available and discoverable. I could create also some docs for it. I guess, you can start preparing your extensions. I could start working on this most likely after the 4.2 release.

Also, this toolkit should have some momentum, I would be ok with breaking changes so we can independently evaluate new Stride functionality and features.

**Reply by Ethereal77** [reply]
If not directly include in its current form, we should discuss how would be the correct way to facilitate those common tasks in a way that fits with the API of the Stride libraries.

**Reply by VaclavElias** [reply]
I agree 🙂

**Comment by programistro** [comment]
in the future, switch Stride to Avalonia for cross-platform=)

**Reply by IXLLEGACYIXL** [reply]
it would be good if such a switch happens to not lose the WYSIWYG editor

the current xamls take forever to load in VS just to throw errors that they cant be displayed

**Reply by Jklawreszuk** [reply]
I think we need to refactor current WPF first and modernize it a little bit. Then separate WPF code from pure logic.

**Reply by IXLLEGACYIXL** [reply]
is it "fix" yet to use avalonia or is  another framework targeted like MAUI ? 
maui doesnt have a WYSIWYG designer at all.. only that stupid hot reload which doesnt really help

**Reply by Doprez** [reply]
MAUI doesn't have proper linux support which is unfortunate. As far as I know all 3 major frameworks have live preview support in VS2022 (Avalonia, Uno and MAUI).

**Reply by tebjan** [reply]
This has been discussed several times and is not relevant to this release. Please continue the discussion on the respective issues, such as:
https://github.com/stride3d/stride/issues/1629 https://github.com/stride3d/stride/issues/360

or on discord or a new discussion thread. so that the argument don't repeat everywhere.

**Reply by programistro** [reply]
You can use the Avalonia SPF to simplify this task

**Reply by programistro** [reply]
I think that WPF is outdated and it's time to switch to another framework, for example Avalonia, the new framework will give great opportunities and flexibility in the UI

**Reply by programistro** [reply]
to begin with, I think it's worth trying to translate the installer to Avalonia

**Comment by NicusorN5** [comment]
Noticed that #1692 and #1700 are caused by #1750 . I suggest replacing the two items with #1750 .

**Reply by VaclavElias** [reply]
Thanks. I updated the summary accordingly.

**Comment by Doprez** [comment]
We can add the following to fix the translation issues found and resolved by @Jklawreszuk 

Nuget for translations
https://github.com/stride3d/gettextnet/pull/2

PR to fix Stride translation
https://github.com/stride3d/stride/pull/1717

Ill let Jklawreszuk confirm if I misunderstood something.

**Reply by Jklawreszuk** [reply]
@Doprez Correct. From what I remember, #1717 itself is complete. Although it would be better to release my gettext fork by @xen2 or If will take too long I can release nuget packages from my fork directly :)

**Reply by VaclavElias** [reply]
I have added those two to the summary at the top :)

**Comment by Doprez** [comment]
Few more things to add for this release
- moving assets automatically https://github.com/stride3d/stride/pull/1827
- Support for F# and VB https://github.com/stride3d/stride/pull/1821
- Dynamic snapping in editor https://github.com/stride3d/stride/pull/1801

Many more bug fixes but these are some new features/QOL changes

**Reply by NicusorN5** [reply]
The PRs you mentioned are already merged. So, obviously they would be added in the next release.

**Reply by Doprez** [reply]
Yes but once the blog post is created for the release it will be a little easier to keep track of features vs bug fixes.

**Reply by VaclavElias** [reply]
While they are available now which is cool, we can highlight these in **4.2** specifically. It is actually nice that we aren't piling things for 4.2, but we are merging them right away to 4.1 so they could be used now.

**Comment by r-bertolini** [comment]
Hi, I would like to ask if there's a roadmap for Stride feature development?
Also, are there any plans to make console release possible?
Thanks!

**Reply by VaclavElias** [reply]
We are preparing some content, which we could call Community Roadmap! Once reviewed, we will add a link here as well.

**Comment by ghost** [comment]
Hmmm great

**Comment by tebjan** [comment]
I've added a few to the list in the post above.

**Comment by Doprez** [comment]
Already fixed thanks to @Jklawreszuk but worth metnioning since some people reported issues with it in discord.

The light probes issue found in 4.2 should be fixed https://github.com/stride3d/stride/pull/2077

**Comment by Doprez** [comment]
We can check off the extension fix, it was merged a few days ago. I am not sure though if @xen2 was able to create the release with all of the fixes after?

**Comment by VaclavElias** [comment]
This is now unpinned.
